File: epi_judge_python/do_lists_overlap.py
import functools
from typing import Optional

from list_node import ListNode
from test_framework import generic_test
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook


def cycle(head):
    its = prev = head
    if its and its.next:
        itf = head.next
        while its and itf:
            if its is itf:
                return its, prev
            prev = its
            its = its.next
            itf = itf.next
            if itf:
                itf = itf.next
    return None, None


def cycle_node(l0, l1):
    n0 = n1 = 0
    it = l0
    while it:
        n0 += 1
        it = it.next
    it = l1
    while it:
        n1 += 1
        it = it.next
    if not n0 or not n1:
        return None
    for _ in range(abs(n0 - n1)):
        if n0 > n1:
            l0 = l0.next
        else:
            l1 = l1.next
    for _ in range(min(n0, n1)):
        if l0 is l1:
            return l0
        l0 = l0.next
        l1 = l1.next
    return None


def overlapping_lists(l0: ListNode, l1: ListNode) -> Optional[ListNode]:
    c0, p0 = cycle(l0)
    c1, p1 = cycle(l1)
    if p0:
        p0.next = None
    if p1:
        p1.next = None
    cn = cycle_node(l0, l1)
    if p0:
        p0.next = c0
    if p1:
        p1.next = c1
    return cn


@enable_executor_hook
def overlapping_lists_wrapper(executor, l0, l1, common, cycle0, cycle1):
    if common:
        if not l0:
            l0 = common
        else:
            it = l0
            while it.next:
                it = it.next
            it.next = common

        if not l1:
            l1 = common
        else:
            it = l1
            while it.next:
                it = it.next
            it.next = common

    if cycle0 != -1 and l0:
        last = l0
        while last.next:
            last = last.next
        it = l0
        for _ in range(cycle0):
            if not it:
                raise RuntimeError("Invalid input data")
            it = it.next
        last.next = it

    if cycle1 != -1 and l1:
        last = l1
        while last.next:
            last = last.next
        it = l1
        for _ in range(cycle1):
            if not it:
                raise RuntimeError("Invalid input data")
            it = it.next
        last.next = it

    common_nodes = set()
    it = common
    while it and id(it) not in common_nodes:
        common_nodes.add(id(it))
        it = it.next

    result = executor.run(functools.partial(overlapping_lists, l0, l1))

    if not (id(result) in common_nodes or (not common_nodes and not result)):
        raise TestFailure("Invalid result")


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "do_lists_overlap.py", "do_lists_overlap.tsv", overlapping_lists_wrapper
        )
    )

File: epi_judge_cpp/do_lists_overlap.cc
#include <set>
#include <stdexcept>

#include "list_node.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"

shared_ptr<ListNode<int>> CycleNode(shared_ptr<ListNode<int>> head) {
  auto slow{head}, fast{head};

  while (fast && fast->next) {
    slow = slow->next;
    fast = fast->next->next;

    if (slow == fast) {
      int clen = 0;
      do {
        clen++;
        slow = slow->next;
      } while (slow != fast);

      auto front = head;
      while (clen--) {
        front = front->next;
      }

      auto back = head;
      while (front != back) {
        front = front->next;
        back = back->next;
      }
      return front;
    }
  }
  while (slow->next) {
    slow = slow->next;
  }
  return slow;
}

shared_ptr<ListNode<int>> OverlappingLists(shared_ptr<ListNode<int>> l0,
                                           shared_ptr<ListNode<int>> l1) {
  if (!l0 || !l1) {
    return nullptr;
  }
  auto cycle0 = CycleNode(l0);
  auto cycle1 = CycleNode(l1);
  if (cycle0->next && cycle1->next) {
    bool overlap = cycle0 == cycle1 ? true : false;
    if (!overlap) {
      for (auto ptr = cycle0->next; ptr != cycle0; ptr = ptr->next) {
        if (ptr == cycle1) {
          overlap = true;
        }
      }
    }
    return overlap ? cycle0 : nullptr;
  }

  auto Count = [](shared_ptr<ListNode<int>> p) {
    int count = 0;
    while (p->next) {
      count++;
      p = p->next;
    }
    return count;
  };

  auto AdvancePtr = [](shared_ptr<ListNode<int>> &p, int count) {
    while (count--) {
      p = p->next;
    }
  };

  if (!cycle0->next && !cycle1->next) {
    if (cycle0 == cycle1) {
      int count0 = Count(l0);
      int count1 = Count(l1);
      AdvancePtr(count0 > count1 ? l0 : l1, abs(count0 - count1));
      while (l0 != l1) {
        l0 = l0->next;
        l1 = l1->next;
      }
      return l0;
    }
  }
  return nullptr;
}
void OverlappingListsWrapper(TimedExecutor &executor,
                             shared_ptr<ListNode<int>> l0,
                             shared_ptr<ListNode<int>> l1,
                             shared_ptr<ListNode<int>> common, int cycle0,
                             int cycle1) {
  if (common) {
    if (!l0) {
      l0 = common;
    } else {
      auto it = l0;
      while (it->next) {
        it = it->next;
      }
      it->next = common;
    }

    if (!l1) {
      l1 = common;
    } else {
      auto it = l1;
      while (it->next) {
        it = it->next;
      }
      it->next = common;
    }
  }

  if (cycle0 != -1 && l0) {
    auto last = l0;
    while (last->next) {
      last = last->next;
    }
    auto it = l0;
    while (cycle0-- > 0) {
      if (!it) {
        throw std::runtime_error("Invalid input data");
      }
      it = it->next;
    }
    last->next = it;
  }

  if (cycle1 != -1 && l1) {
    auto last = l1;
    while (last->next) {
      last = last->next;
    }
    auto it = l1;
    while (cycle1-- > 0) {
      if (!it) {
        throw std::runtime_error("Invalid input data");
      }
      it = it->next;
    }
    last->next = it;
  }

  std::set<shared_ptr<ListNode<int>>> common_nodes;
  auto it = common;
  while (it && common_nodes.count(it) == 0) {
    common_nodes.insert(it);
    it = it->next;
  }

  auto result = executor.Run([&] { return OverlappingLists(l0, l1); });

  if (!((common_nodes.empty() && result == nullptr) ||
        common_nodes.count(result) > 0)) {
    throw TestFailure("Invalid result");
  }
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "l0",     "l1",
                                       "common",   "cycle0", "cycle1"};
  return GenericTestMain(args, "do_lists_overlap.cc", "do_lists_overlap.tsv",
                         &OverlappingListsWrapper, DefaultComparator{},
                         param_names);
}
